addon/wait-for-promise.ts

```diff
@@ -23,24 +24,83 @@ const PROMISE_WAITER = new TestWaiter<Promise<unknown>>('promise-waiter');
  *     }));
  *   }
  * }
+ *
+ * @example
+ *
+ * import Component from '@ember/component';
+ * import { waitForPromise } from 'ember-test-waiters';
+ *
+ * export default class Friendz extends Component {
+ *   @waitForPromise
+ *   async doAsyncStuff() {
+ *     await somethingAsync();
+ *   }
+ * }
+ *
+ * @example
+ *
+ * import Component from '@ember/component';
+ * import { waitForPromise } from 'ember-test-waiters';
+ *
+ * export default Component.extend({
+ *   doAsyncStuff: waitForPromise(async function doAsyncStuff() {
+ *     await somethingAsync();
+ *   }
+ * });
  */
-export default function waitForPromise<T>(promise: Promise<T>, label?: string) {
-  let result = promise;
+export default function waitForPromise(
+  target: object,
+  _key: string,
+  descriptor: PropertyDescriptor,
+  label?: string
+): PropertyDescriptor;
+export default function waitForPromise<T>(promise: Promise<T>, label?: string): Promise<T>;
+export default function waitForPromise<T>(fn: Function, label?: string): Function;
+
+export default function waitForPromise<T>(
+  ...args:
+    | [object, string, PropertyDescriptor, string?]
+    | [Promise<T>, string?]
+    | [Function, string?]
+): PropertyDescriptor | Promise<T> | Function {
+  if (args.length < 3) {
```

Comment:
Similar to my comment above, can we extract this condition into a variable that makes it clear that this is the functional invocation branch of the code?